File: mcp-code-intelligence-nodejs/src/memory/tests/working-tier-expiry.test.ts
/**
 * Unit tests for WorkingTierExpiry — lazy auto-expiry of stale entries.
 */

import { describe, it, beforeEach } from 'node:test';
import assert from 'node:assert/strict';
import Database from 'better-sqlite3';
import { WorkingTierExpiry } from '../v2/working-tier-expiry.js';

describe('WorkingTierExpiry', () => {
  let db: Database.Database;
  let expiry: WorkingTierExpiry;

  beforeEach(() => {
    db = new Database(':memory:');
    db.exec(`
      CREATE TABLE knowledge_entries (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        content TEXT NOT NULL,
        tier TEXT NOT NULL DEFAULT 'WORKING',
        pinned INTEGER NOT NULL DEFAULT 0,
        archived INTEGER NOT NULL DEFAULT 0,
        quality_score INTEGER DEFAULT NULL,
        created_at TEXT NOT NULL DEFAULT (datetime('now')),
        updated_at TEXT NOT NULL DEFAULT (datetime('now'))
      )
    `);
    // Use 1 hour expiry for testing
    expiry = new WorkingTierExpiry(db, { expiryHours: 1, promoteThreshold: 60 });
  });

  function insertEntry(opts: {
    tier?: string; pinned?: number; archived?: number;
    quality_score?: number | null; hoursAgo?: number;
  }): number {
    const createdAt = new Date(Date.now() - (opts.hoursAgo ?? 0) * 3600_000).toISOString();
    const result = db.prepare(`
      INSERT INTO knowledge_entries (content, tier, pinned, archived, quality_score, created_at)
      VALUES ('test content', ?, ?, ?, ?, ?)
    `).run(
      opts.tier ?? 'WORKING',
      opts.pinned ?? 0,
      opts.archived ?? 0,
      opts.quality_score ?? null,
      createdAt
    );
    return Number(result.lastInsertRowid);
  }

  describe('processStale', () => {
    it('returns empty when no stale entries', () => {
      insertEntry({ hoursAgo: 0 }); // fresh entry
      const actions = expiry.processStale();
      assert.equal(actions.length, 0);
    });

    it('promotes high-quality stale entries to EPISODIC', () => {
      const id = insertEntry({ hoursAgo: 2, quality_score: 75 });
      const actions = expiry.processStale();
      assert.equal(actions.length, 1);
      assert.deepEqual(actions[0], {
        entry_id: id, action: 'promoted', quality_score: 75, to_tier: 'EPISODIC',
      });
      // Verify DB updated
      const row = db.prepare('SELECT tier FROM knowledge_entries WHERE id = ?').get(id) as any;
      assert.equal(row.tier, 'EPISODIC');
    });

    it('archives low-quality stale entries', () => {
      const id = insertEntry({ hoursAgo: 2, quality_score: 30 });
      const actions = expiry.processStale();
      assert.equal(actions.length, 1);
      assert.deepEqual(actions[0], {
        entry_id: id, action: 'archived', quality_score: 30,
      });
      const row = db.prepare('SELECT archived FROM knowledge_entries WHERE id = ?').get(id) as any;
      assert.equal(row.archived, 1);
    });

    it('archives entries with null quality_score (treated as 0)', () => {
      insertEntry({ hoursAgo: 2, quality_score: null });
      const actions = expiry.processStale();
      assert.equal(actions[0].action, 'archived');
    });

    it('skips pinned entries (BR-F1-05)', () => {
      insertEntry({ hoursAgo: 2, quality_score: 30, pinned: 1 });
      const actions = expiry.processStale();
      assert.equal(actions.length, 0);
    });

    it('skips already archived entries', () => {
      insertEntry({ hoursAgo: 2, quality_score: 30, archived: 1 });
      const actions = expiry.processStale();
      assert.equal(actions.length, 0);
    });

    it('skips non-WORKING tier entries', () => {
      insertEntry({ hoursAgo: 2, quality_score: 30, tier: 'EPISODIC' });
      const actions = expiry.processStale();
      assert.equal(actions.length, 0);
    });

    it('processes multiple entries in one call', () => {
      insertEntry({ hoursAgo: 2, quality_score: 80 }); // promote
      insertEntry({ hoursAgo: 3, quality_score: 20 }); // archive
      insertEntry({ hoursAgo: 4, quality_score: 60 }); // promote (exactly at threshold)
      const actions = expiry.processStale();
      assert.equal(actions.length, 3);
      assert.equal(actions.filter(a => a.action === 'promoted').length, 2);
      assert.equal(actions.filter(a => a.action === 'archived').length, 1);
    });
  });
});

File: mcp-code-intelligence-kotlin/src/test/kotlin/com/codeintel/memory/search/WorkingTierExpiryTest.kt
/** Unit tests for WorkingTierExpiry — lazy auto-expiry of stale entries. */
package com.codeintel.memory.search

import java.sql.DriverManager
import java.time.Instant
import java.time.temporal.ChronoUnit
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class WorkingTierExpiryTest {

    private fun createDb(): java.sql.Connection {
        val conn = DriverManager.getConnection("jdbc:sqlite::memory:")
        conn.createStatement().execute("""
            CREATE TABLE knowledge_entries (
              id INTEGER PRIMARY KEY AUTOINCREMENT,
              content TEXT NOT NULL,
              tier TEXT NOT NULL DEFAULT 'WORKING',
              archived INTEGER NOT NULL DEFAULT 0,
              pinned INTEGER NOT NULL DEFAULT 0,
              quality_score INTEGER DEFAULT NULL,
              created_at TEXT NOT NULL DEFAULT (datetime('now')),
              updated_at TEXT NOT NULL DEFAULT (datetime('now'))
            )
        """.trimIndent())
        return conn
    }

    private fun insertEntry(
        conn: java.sql.Connection,
        tier: String = "WORKING",
        archived: Int = 0,
        pinned: Int = 0,
        qualityScore: Int? = null,
        hoursAgo: Long = 48
    ): Long {
        val createdAt = Instant.now().minus(hoursAgo, ChronoUnit.HOURS).toString()
        conn.prepareStatement(
            "INSERT INTO knowledge_entries (content, tier, archived, pinned, quality_score, created_at) VALUES (?, ?, ?, ?, ?, ?)"
        ).use { stmt ->
            stmt.setString(1, "test content")
            stmt.setString(2, tier)
            stmt.setInt(3, archived)
            stmt.setInt(4, pinned)
            if (qualityScore != null) stmt.setInt(5, qualityScore) else stmt.setNull(5, java.sql.Types.INTEGER)
            stmt.setString(6, createdAt)
            stmt.executeUpdate()
        }
        return conn.createStatement().executeQuery("SELECT last_insert_rowid()").use { rs ->
            rs.next(); rs.getLong(1)
        }
    }

    @Test
    fun `promotes high quality stale entries to EPISODIC`() {
        val conn = createDb()
        insertEntry(conn, qualityScore = 75, hoursAgo = 48)
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertEquals(1, actions.size)
        assertEquals("promoted", actions[0].action)
        assertEquals("EPISODIC", actions[0].toTier)
    }

    @Test
    fun `archives low quality stale entries`() {
        val conn = createDb()
        insertEntry(conn, qualityScore = 20, hoursAgo = 48)
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertEquals(1, actions.size)
        assertEquals("archived", actions[0].action)
    }

    @Test
    fun `skips pinned entries`() {
        val conn = createDb()
        insertEntry(conn, pinned = 1, qualityScore = 10, hoursAgo = 48)
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertTrue(actions.isEmpty())
    }

    @Test
    fun `skips non-WORKING tier entries`() {
        val conn = createDb()
        insertEntry(conn, tier = "EPISODIC", qualityScore = 10, hoursAgo = 48)
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertTrue(actions.isEmpty())
    }

    @Test
    fun `skips already archived entries`() {
        val conn = createDb()
        insertEntry(conn, archived = 1, qualityScore = 10, hoursAgo = 48)
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertTrue(actions.isEmpty())
    }

    @Test
    fun `skips entries newer than expiryHours`() {
        val conn = createDb()
        insertEntry(conn, qualityScore = 10, hoursAgo = 2) // only 2 hours old
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertTrue(actions.isEmpty())
    }

    @Test
    fun `processes multiple entries in one call`() {
        val conn = createDb()
        insertEntry(conn, qualityScore = 80, hoursAgo = 48) // promote
        insertEntry(conn, qualityScore = 15, hoursAgo = 48) // archive
        insertEntry(conn, qualityScore = 65, hoursAgo = 48) // promote
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertEquals(3, actions.size)
        assertEquals(2, actions.count { it.action == "promoted" })
        assertEquals(1, actions.count { it.action == "archived" })
    }

    @Test
    fun `null quality score treated as 0`() {
        val conn = createDb()
        insertEntry(conn, qualityScore = null, hoursAgo = 48)
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertEquals(1, actions.size)
        assertEquals("archived", actions[0].action)
    }

    @Test
    fun `returns empty list when no stale entries`() {
        val conn = createDb()
        val expiry = WorkingTierExpiry(conn, expiryHours = 24)
        val actions = expiry.processStale()
        assertTrue(actions.isEmpty())
    }
}
